public/app/plugins/datasource/tempo/QueryEditor/NativeSearch.tsx

```diff
@@ -48,60 +48,70 @@ const NativeSearch = ({ datasource, query, onChange, onBlur, onRunQuery }: Props
   const styles = useStyles2(getStyles);
   const languageProvider = useMemo(() => new TempoLanguageProvider(datasource), [datasource]);
   const [hasSyntaxLoaded, setHasSyntaxLoaded] = useState(false);
-  const [autocomplete, setAutocomplete] = useState<{
-    serviceNameOptions: Array<SelectableValue<string>>;
-    spanNameOptions: Array<SelectableValue<string>>;
-  }>({
-    serviceNameOptions: [],
-    spanNameOptions: [],
+  const [asyncServiceNameValue, setAsyncServiceNameValue] = useState<SelectableValue<any>>({
+    value: '',
+  });
+  const [asyncSpanNameValue, setAsyncSpanNameValue] = useState<SelectableValue<any>>({
+    value: '',
   });
   const [error, setError] = useState(null);
   const [inputErrors, setInputErrors] = useState<{ [key: string]: boolean }>({});
+  const [isLoading, setIsLoading] = useState<{
+    serviceName: boolean;
+    operationName: boolean;
+  }>({
+    serviceName: false,
+    operationName: false,
+  });
 
-  const fetchServiceNameOptions = useMemo(
-    () =>
-      debounce(
-        async () => {
-          const res = await languageProvider.getOptions('service.name');
-          setAutocomplete((prev) => ({ ...prev, serviceNameOptions: res }));
-        },
-        500,
-        { leading: true, trailing: true }
-      ),
+  async function loadOptionsAsync(nameType: string, lp: TempoLanguageProvider) {
+    const res = await lp.getOptions(nameType === 'serviceName' ? 'service.name' : 'name');
+    setIsLoading({
+      serviceName: nameType === 'serviceName' && false,
```

Comment:
Once you have `spanName` as a key in `isLoading`, you can update the loading state with this:
`setIsLoading({ ...isLoading, [nameType]: false })` which I think will be more readable than the conditionals here